### PR TITLE
Support list<Foo>

### PR DESCRIPTION
**Pull request by orklah**
As discussed in phpstan/phpstan/issues/2788, this PR implements the notations
`list` and `list<Foo>` as aliases of respectively `array<int, mixed>` and `array<int, Foo>`

I added some tests on ReturnTypeRuleTest.php but I'm not sure it was the best place to put them. Please tell me if it shoud be located elsewhere or if it's missing tests.

Also, should I try to add some tests in https://github.com/phpstan/phpdoc-parser to ensure those notations will stay stable?

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Type inference is best tested like this (the part with `tests/PHPStan/Analyser/NodeScopeResolverTest.php`): https://github.com/phpstan/phpstan-src/commit/65e578c8fd8c3628aa022705ac67511846c37f17#diff-b5a18b99fdc03e387e605f09db47dc15

**Comment by ondrejmirtes** [comment]
> Also, should I try to add some tests in https://github.com/phpstan/phpdoc-parser to ensure those notations will stay stable?

No, isn't necessary.

File: tests/PHPStan/Analyser/data/list-type.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace ListType;
+
+use function PHPStan\Analyser\assertType;
+
+class Foo
+{
+	public function withoutGenerics(): void
+	{
+		/** @var list $list */
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/list-type.php]: You should at least once typehint the type as method parameter and assert it right at the beginning of the body. If you assert the variable only after some changes (array appending), you have no guarantees about the state right after `$list = [];`.

**Comment by ondrejmirtes** [comment]
Otherwise perfect! Thank you.

File: tests/PHPStan/Analyser/data/list-type.php (hunk 1 above)
Comment (orklah) [reply tests/PHPStan/Analyser/data/list-type.php]: Ok, I added some test cases with direct assertions. It should be ok

**Comment by ondrejmirtes** [comment]
Please get rid of the merge commits and rebase it onto master. I'm not able to wrap my head around this Git history. Thank you.

**Comment by ondrejmirtes** [comment]
It'd probably be easiest to start a new branch and just do all the changes in a single commit.

**Comment by orklah** [comment]
> It'd probably be easiest to start a new branch and just do all the changes in a single commit.

Yeah, I'll probably do that. I'm not really an expert with git!

**Pull request closed without merging**